waspc/src/Analyzer/Evaluator/Combinators.hs

```diff
@@ -0,0 +1,175 @@
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+
+-- | This module contains combinators for building evaluators to convert
+-- typed Wasp AST into Haskell values.
+--
+-- A typical use:
+--
+-- @
+-- data Page = Page { title :: String, author :: Maybe String, content :: String }
+--
+-- page :: Evaluator Page
+-- page = dict $ Page <$> field "title" string <*> maybeField "author" string <*> field "content" string
+-- @
+--
+-- This creates a "Evaluator" @page@ that would turn the Wasp expression
+-- @{ title: "Home", content: "Hello world" }@ into
+-- @Page { title = "Home", author = Nothing, content = "Hello world" }@
+module Analyzer.Evaluator.Combinators
+  ( -- * Types
+    Evaluator,
+    DictEvaluator,
+
+    -- * Functions
+    runEvaluator,
+
+    -- * "Evaluator" combinators
+    string,
+    integer,
+    double,
+    bool,
+    decl,
+    enum,
+    dict,
+    list,
+    extImport,
+    json,
+    psl,
+
+    -- * "DictEvaluator" combinators
+    field,
+    maybeField,
+  )
+where
+
+import Analyzer.Evaluator.Decl
+import Analyzer.Evaluator.EvaluationError
+import qualified Analyzer.Evaluator.Types as E
+import qualified Analyzer.Type as T
+import Analyzer.TypeChecker.AST (TypedExpr (..), exprType)
+import qualified Analyzer.TypeDefinitions as TD
+import Control.Arrow (left)
+import Data.Functor.Compose (Compose (Compose, getCompose))
+import qualified Data.HashMap.Strict as H
+import Data.Typeable (cast)
+
+-- | Bindings for currently evaluated declarations
+type Bindings = H.HashMap String Decl
+
+-- | The context in an evaluation.
+type EvalCtx a = (TD.TypeDefinitions, Bindings, a)
+
+-- | An evaluation from "a" to "b" with the evaluation context.
+type (|>) a b = Compose ((->) (EvalCtx a)) (Either EvaluationError) b
```

Comment:
I am trying to make this change myself right now to see if it even works, will let you know how it goes.

Reply:
I tried it out and it seems to work great! I made a PR with it, you can take a look here: https://github.com/wasp-lang/wasp/pull/301 . It is aimed to this branch, so once we merge it it will directly update this PR here.

Reply:
I was additionally wondering if we can somehow remove duplication between Evaluator and DictEvaluator, so I made additional change where I extracted common functionality and made that as an additional commit on #301. So now #301 has two commits -> first separates context from input, second one removes duplication between Evaluator and DictEvaluator. Check it out, we can merge only specific commits if that will be needed.

Reply:
Hm I just realized that with second commit we loose some of the type safety that newtype gives us. Not sure if that is a problem or not. In theory, somebody could combine the combinators wrong I guess. Not right now, but in the future if some combinators work with two different evaluators that have same shape but are really different, then that could be an issue.

Reply:
Done